partial fix for buffer mutations with aot_compile

bdhirsh opened the pull request:
Partial fix for https://github.com/pytorch/pytorch/issues/120424. @int3 to continue investigation.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #120828
* __->__ #120427



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames

bdhirsh committed:
partial fix for buffer mutations with aot_compile

[ghstack-poisoned]

bdhirsh commented on torch/_inductor/compile_fx.py:
```diff
@@ -189,7 +189,7 @@ def _unlift_graph(mod, gm, graph_signature):
     mutated_outputs = []
     for out in outputs:
-        if out in graph_signature.buffers_to_mutate:
+        if str(out) in graph_signature.buffers_to_mutate:
```
cc @angelayi - do you happen to know why inductor has its own unlift()` function? One thing I was wondering if is this code could reasonable be shared with `ExportedProgram.module()`.

I think this part was slightly broken, because `graph_signature.buffers_to_mutate` contains a bunch of strings (the name of each node), but `out` here is the fx node itself.

aakhundov commented on torch/_inductor/ir.py:
```diff
@@ -6590,7 +6590,10 @@ def create(data):
-            return self.data.get_name() in V.graph.graph_inputs
+            if self.data.get_name() in V.graph.graph_inputs:
+                return True
+        if isinstance(self.data, ConstantBuffer):
```
```suggestion
        elif isinstance(self.data, ConstantBuffer):
```

ezyang commented:
Is @int3 gonna commandeer this? Needs test

int3 commented:
hm I'm gonna write a fix for inductor that depends on this fix, but I think we have two separate issues here. @bdhirsh could you write a test for the expected fx graph output that this PR fixes?

int3 committed:
Update on "partial fix for buffer mutations with aot_compile"

Partial fix for https://github.com/pytorch/pytorch/issues/120424. int3 to continue investigation.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

bdhirsh commented:
This PR is subsumed by https://github.com/pytorch/pytorch/pull/121688
